File: PiTracerLib/PiTracerLib/Scene.cs
// This file is part of the ArmoniK project
//
// Copyright (C) ANEO, 2021-$CURRENT_YEAR$. All rights reserved.
//   W. Kirschenmann   <wkirschenmann@aneo.fr>
//   J. Gurhem         <jgurhem@aneo.fr>
//   D. Dubuc          <ddubuc@aneo.fr>
//   L. Ziane Khodja   <lzianekhodja@aneo.fr>
//   F. Lemaitre       <flemaitre@aneo.fr>
//   S. Djebbar        <sdjebbar@aneo.fr>
//   J. Fonseca        <jfonseca@aneo.fr>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

namespace PiTracerLib
{
  public class Scene
  {
    public int        ImgWidth   { get; }
    public int        ImgHeight  { get; }
    public int        KillDepth  { get; }
    public int        SplitDepth { get; }
    public Camera     Camera     { get; }
    public SphereList Spheres    { get; }

    public Scene(byte[] payload)
    {
      ImgWidth = BitConverter.ToInt32(payload, 0);
      ImgHeight = BitConverter.ToInt32(payload, 4);
      KillDepth = BitConverter.ToInt32(payload, 8);
      SplitDepth = BitConverter.ToInt32(payload, 12);
      var offset = 12 + 4;
      Camera = new Camera(payload, ImgWidth, ImgHeight, offset);
      offset += Camera.Size;
      var spheres = new Sphere[(payload.Length - offset) / Sphere.Size];
      for (var i = 0; i < spheres.Length; i++)
      {
        spheres[i] = new Sphere(payload, offset + Sphere.Size * i);
      }

      Spheres = new SphereList(spheres);
    }

    public Scene(int imgWidth,
                         int imgHeight,
                         int killDepth,
                         int splitDepth,
                         Camera camera,
                         Sphere[] spheres)
    {
      ImgWidth = imgWidth;
      ImgHeight = imgHeight;
      KillDepth = killDepth;
      SplitDepth = splitDepth;
      Camera = camera;
      Spheres = new SphereList(spheres);
    }

    public byte[] ToBytes()
    {
      var bytes = new byte[36 + Camera.Size + Spheres.Length * Sphere.Size];
      BitConverter.GetBytes(ImgWidth).CopyTo(bytes, 0);
      BitConverter.GetBytes(ImgHeight).CopyTo(bytes, 4);
      BitConverter.GetBytes(KillDepth).CopyTo(bytes, 8);
      BitConverter.GetBytes(SplitDepth).CopyTo(bytes, 12);
      Camera.ToBytes().CopyTo(bytes, 16);
      for (var i = 0; i < Spheres.Length; i++)
      {
        Spheres[i].ToBytes().CopyTo(bytes, 16 + Camera.Size + i * Sphere.Size);
      }

      return bytes;
    }
  }
}

File: Client/tracer/scene.py
from tracer.objects import Sphere, Reflection, Camera, Scene

spheres = [
    Sphere(
        1e5,
        [1e5 - 5, 40.8, 81.6],
        [0.0, 0.0, 0.0],
        [0.75, 0.25, 0.25],
        Reflection.DIFF,
        -1,
    ),
    Sphere(
        1e5,
        [-1e5 + 104, 40.8, 81.6],
        [0.0, 0.0, 0.0],
        [0.25, 0.25, 0.75],
        Reflection.DIFF,
        -1,
    ),
    Sphere(
        1e5, [50, 40.8, 1e5], [0.0, 0.0, 0.0], [0.75, 0.75, 0.75], Reflection.DIFF, -1
    ),
    Sphere(
        1e5,
        [50, 40.8, -1e5 + 170],
        [0.0, 0.0, 0.0],
        [0.0, 0.0, 0.0],
        Reflection.DIFF,
        -1,
    ),
    Sphere(
        1e5, [50, 1e5, 81.6], [0.0, 0.0, 0.0], [0.75, 0.75, 0.75], Reflection.DIFF, -1
    ),
    Sphere(
        1e5,
        [50, -1e5 + 81.6, 81.6],
        [0.0, 0.0, 0.0],
        [0.75, 0.75, 0.75],
        Reflection.DIFF,
        -1,
    ),
    Sphere(
        16.5,
        [40, 16.5, 47],
        [0.0, 0.0, 0.0],
        [0.999, 0.999, 0.999],
        Reflection.SPEC,
        -1,
    ),
    Sphere(
        10, [90, 25, 125], [0.0, 0.0, 0.0], [0.999, 0.999, 0.999], Reflection.SPEC, -1
    ),
    Sphere(
        16.5,
        [73, 46.5, 94],
        [0.0, 0.0, 0.0],
        [0.999, 0.999, 0.999],
        Reflection.REFR,
        -1,
    ),
    Sphere(
        10, [15, 45, 112], [0.0, 0.0, 0.0], [0.999, 0.999, 0.999], Reflection.DIFF, -1
    ),
    Sphere(15, [16, 16, 130], [0.0, 0.0, 0.0], [0, 0.999, 0], Reflection.REFR, -1),
    Sphere(7.5, [40, 8, 120], [0.0, 0.0, 0.0], [0.999, 0.999, 0], Reflection.REFR, -1),
    Sphere(8.5, [67, 9, 122], [0.1, 0.1, 0.0], [0.999, 0.999, 0], Reflection.REFR, -1),
    Sphere(10, [80, 12, 92], [0.1, 1, 0.1], [0.1, 0.6, 0.1], Reflection.DIFF, -1),
    Sphere(
        600, [50, 681.33, 81.6], [1.5, 1.5, 1.5], [0.0, 0.0, 0.0], Reflection.DIFF, -1
    ),
    Sphere(9, [95, 65, 81.6], [0.0, 0.4, 0.8], [0.1, 0.3, 0.6], Reflection.DIFF, -1),
    Sphere(8, [15, 70, 75], [1, 0.1, 0.1], [0.6, 0.1, 0.1], Reflection.DIFF, -1),
]

camera = Camera()


def get_scene(
    img_width: int, img_height: int, kill_depth: int, split_depth: int
) -> Scene:
    return Scene(img_width, img_height, kill_depth, split_depth, camera, spheres)
